File: PSView2/ProcessFriendlyName.cpp
#include "stdafx.h"

namespace PSView2
{
	/*****************************************************
	* class ProcessFriendlyName implementation
	*****************************************************/
	ProcessFriendlyName::ProcessFriendlyName() 	// Protected
	{
	}

	ProcessFriendlyName::ProcessFriendlyName(String^ name) 	// Service
	{
		szName = name->Trim();
		fVisible = true;
		dtDate = DateTime::FromFileTimeUtc(0);
	}

	ProcessFriendlyName::ProcessFriendlyName(String^ name, DateTime date, bool visible)	// Window
	{
		szName = name->Trim();
		dtDate = date;
		fVisible = visible;
	}

	/* The ProcessFriendlyName can be sorted and these are the criteria:
	Visibility first
	Date of window creation
	Then the string name
	*/
	int ProcessFriendlyName::CompareTo(Object^ obj)
	{
		ProcessFriendlyName^ rhs = dynamic_cast<ProcessFriendlyName^>(obj);
		if (op_equality(this, rhs))
			return 0;
		if (op_lessthan(this, rhs))
			return -1;
		return 1;
	}

	bool ProcessFriendlyName::op_equality(ProcessFriendlyName^ _j, ProcessFriendlyName^ _k)
	{
		if (_j->Visible != _k->Visible)
			return false;
		if (_j->Date != _k->Date)
			return false;
		if (String::Compare(_j->Name, _k->Name) != 0)
			return false;
		return true;
	}

	bool ProcessFriendlyName::op_lessthan(ProcessFriendlyName^ _j, ProcessFriendlyName^ _k)
	{
		if (_j->Visible == true && _k->Visible == false)
			return true;
		if (_j->Visible == false && _k->Visible == true)
			return false;
		if (_j->Date < _k->Date)
			return true;
		if (_j->Date > _k->Date)
			return false;
		if (String::Compare(_j->Name, _k->Name) < 0)
			return true;
		return false;
	}
}

File: ProcessTools/ProcessFriendlyName.cs
using System;
using System.Collections.Generic;

namespace ProcessTools
{
    /// <summary>
    /// This class describes a set of properties for an associated friendly name for a process.
    /// A process can have more than one friendly name whether it be an open window or a service.
    /// The Date variable is used in the window case to determine the first window created by
    /// an application to help sort out the primary window from derivative windows.
    /// </summary>
    public class ProcessFriendlyName : IComparable<ProcessFriendlyName>, IComparer<ProcessFriendlyName>
    {
        /// <summary>
        /// The name of the window or service in this process.
        /// </summary>
        public string Name { get; private set; }
        /// <summary>
        /// The date the window was created (not relevant for services).
        /// </summary>
        public DateTime Date { get; private set; }
        /// <summary>
        /// Whether the window is visible (not relevant for services).
        /// </summary>
        public bool Visible { get; private set; }

        /// <summary>
        /// Service name
        /// </summary>
        public ProcessFriendlyName(string name)
        {
            Name = name.Trim();
            Visible = true;
            Date = DateTime.MinValue;
        }

        /// <summary>
        /// Window name
        /// </summary>
        public ProcessFriendlyName(string name, DateTime date, bool visible)
        {
            Name = name.Trim();
            Date = date;
            Visible = visible;
        }

        /// <summary>
        /// The ProcessFriendlyName can be sorted and these are the criteria:
        ///   Visibility first
        ///   Date of window creation 
        ///   Then the string name
        /// </summary>
        public int CompareTo(ProcessFriendlyName? other)
        {
            if (other == null)
            {
                return -1;
            }
            int compare = -Visible.CompareTo(other.Visible);
            if (compare != 0)
            {
                return compare;
            }
            compare = Date.CompareTo(other.Date);
            if (compare != 0)
            {
                return compare;
            }
            return string.Compare(Name, other.Name);
        }

        /// <summary>
        /// The ProcessFriendlyName can be sorted and these are the criteria:
        ///   Visibility first
        ///   Date of window creation 
        ///   Then the string name
        /// </summary>
        public int Compare(ProcessFriendlyName? x, ProcessFriendlyName? y)
        {
            if(x == null && y == null)
            {
                return 0;
            }
            if(x == null)
            {
                return -1;
            }
            if(y == null)
            {
                return 1;
            }
            return x.CompareTo(y);
        }
    }
}